packages/devextreme/testing/tests/DevExpress.ui.widgets/tabs.tests.js

```diff
@@ -422,6 +422,100 @@ QUnit.module('General', () => {
 });
 
 QUnit.module('Tab select action', () => {
+    QUnit.module('onSelectionChanging', {
+        beforeEach() {
+            this.onSelectionChangingSpy = sinon.spy();
+            this.onSelectionChangedSpy = sinon.spy();
+            this.clock = sinon.useFakeTimers();
+
+            this.$tabs = $('#tabs').dxTabs({
+                items: [
+                    { text: '0' },
+                    { text: '1' },
+                    { text: '2' },
+                    { text: '3' }
+                ],
+                onSelectionChanging: this.onSelectionChangingSpy,
+                onSelectionChanged: this.onSelectionChangedSpy,
+                selectedIndex: 0
+            });
+            this.tabInstance = this.$tabs.dxTabs('instance');
+        },
+        afterEach() {
+            this.clock.restore();
+        }
+    }, () => {
+        QUnit.test('cancelling selection event synchronously', function(assert) {
+            this.onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = true;
+            });
+
+            this.tabInstance.option('onSelectionChanging', this.onSelectionChangingSpy);
+            this.$item = this.$tabs.find('.dx-tab').eq(1);
+            this.$item.trigger('dxclick');
+
+            assert.strictEqual(this.onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should be called');
+            assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called');
+            assert.ok(this.onSelectionChangingSpy.getCall(0).args[0].cancel, 'e.cancel should be set to true');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 when cancelled');
```

Comment:
[test coverage]
There a lot of properties related to selection, not only `selectedIndex`: `selectedItem`, `selectedItems`, `selectedItemKeys`. I prefer we add asserts for all these properties to ensure they are well synchronized.

Reply:
[48e4ac1](https://github.com/DevExpress/DevExtreme/pull/27860/commits/48e4ac14372f22296446a8d3f5ca0147a58bb4b2)